File: src/MaxCounters.java
import java.util.Arrays;

public class MaxCounters {

    public int[] solution(int N, int[] A){
        int max = 0, applyMax = 0;;
        int[] result = new int[N];

        for (int i = 0; i < A.length; ++i) {
            int a = A[i];

            if (a == N + 1) {
                applyMax = max;
            }

            if (1 <= a && a <= N) {
                result[A[i] - 1] = Math.max(applyMax, result[A[i] - 1]);
                max = Math.max(max, ++result[A[i] - 1]);
            }
        }

        for (int i = 0; i < N; ++i) {
            if (result[i] < applyMax) {
                result[i] = applyMax;
            }
        }

        return result;
    }
    public static void main(String args[]) {
        MaxCounters MX = new MaxCounters();
        System.out.println(Arrays.toString(MX.solution(5, new int[]{3, 1, 2, 4, 3, 6, 1})));
    }
}

File: src/JS/MaxCounters.js
function MaxCounters(N, A) {
  let counters = new Array(N).fill(0);
  let maxCount = 0;
  let applyMax = 0;

  for (let i = 0; i < A.length; i++) {
    if (A[i] <= N) {
      counters[A[i] - 1] = Math.max(counters[A[i] - 1], applyMax);
      counters[A[i] - 1]++;
      maxCount = Math.max(maxCount, counters[A[i] - 1]);
    } else {
      applyMax = maxCount;
    }
  }

  for (let i = 0; i < N; i++) {
    counters[i] = Math.max(counters[i], applyMax);
  }
  return counters;
}

console.log(MaxCounters(5, [3, 4, 4, 6, 1, 4, 4]));
